### PR TITLE
Initial commit of API adapter

### LINKED ISSUE #1509: Android - Implement network / caching binding to JS

### PR DESCRIPTION
**Pull request by thisandagain**
Using an S3 bucket to mock the API server right now, but this adds basic network request and caching functionality to the app. Resolves #1509

### CODE REVIEW

**Comment by thisandagain** [comment]
@k88hudson R?

**Comment by thisandagain** [comment]
cc @cadecairos

File: www_src/lib/api.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  options.uri = 'https://gist.githubusercontent.com/thisandagain' + options.uri;
+
+  // Set cache key
+  var key = 'cache::' + options.method.toLowerCase() + '::' + options.uri;
+  
+  // Use device cache if window.Android is available & options.useCache is true
+  if (window.Android && options.useCache === true) {
+    window.Android.logText('Fetching from cache "' + key + '"');
+    var hit = window.Android.getSharedPreferences(key);
+    if (typeof hit !== 'undefined') return callback(null, JSON.parse(hit));
```
Comment (k88hudson) [inline www_src/lib/api.js]: hit is either going to be `null` or a string here, not `undefined`

**Comment by thisandagain** [comment]
Good catch on the cache hit check. Fixed.

**Comment by k88hudson** [comment]
Looks good to me. This is surprisingly simple!

File: www_src/lib/api.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  
+  // Use device cache if window.Android is available & options.useCache is true
+  if (window.Android && options.useCache === true) {
+    window.Android.logText('Fetching from cache "' + key + '"');
+    var hit = window.Android.getSharedPreferences(key);
+    if (typeof hit === 'string') return callback(null, JSON.parse(hit));
+  }
+
+  // XHR request
+  xhr(options, function (err, res, body) {
```
Comment (k88hudson) [inline www_src/lib/api.js]: you can use an arrow function here if you like

**Comment by thisandagain** [comment]
Woo! :smile:

File: www_src/lib/api.js (hunk 2 above)
Comment (thisandagain) [inline www_src/lib/api.js]: Will you do an ES6 for n00bs workshop? My brain still hasn't adapted.

**Pull request merged**